### PR TITLE
WIP: Move all monkey-patched built-ins into single object

### PR DESCRIPTION
**Pull request by dotsdl**
We want all `__xonsh_*` builtins to be moved into a single `__xonsh` object, grouping them all together in a single object for sanity's sake. This will require changes throughout the core library, and probably changes in downstream packages that access these builtins directly.

### CODE REVIEW

**Comment by scopatz** [comment]
Thanks for putting this n @dotsdl! You may need to merge from master now, in order to get the changes from the sprints

**Comment by dotsdl** [comment]
@scopatz no problem! Still have a ways to go but I'll keep plugging away at it.

**Comment by scopatz** [comment]
Thanks @dotsdl! This will be great to have!

**Comment by dotsdl** [comment]
I need to merge changes from master yet; this was an extensive set of changes touching every file, so bracing for this. Need to do it sooner than later though.

**Comment by dotsdl** [comment]
Could use some help here. At this point `xonsh` can start up, but it can't access any of the stuff under `__xonsh__`, such as `__xonsh__.env`. The limited traceback I get looks like:

```
> xonsh --no-rc
> $SOMETHING = 2
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/home/alter/Library/xonsh/xonsh/base_shell.py", line 330, in default
    run_compiled_code(code, self.ctx, None, 'single')
  File "/home/alter/Library/xonsh/xonsh/codecache.py", line 68, in run_compiled_code
    func(code, glb, loc)
  File "<xonsh-code>", line 1, in <module>
NameError: name '__xonsh__.env' is not defined
```
Is this a hard-to-spot use of `getattr(builtins, var)` somewhere, or something more subtle?

**Comment by scopatz** [comment]
I think that the AST is getting generated incorrectly.  I'll try to push up a fix

**Comment by scopatz** [comment]
OK @dotsdl! I have pushed up a fix to your branch for that issue that lets xonsh run! The tests still need some work, and there are some unfortuate merge conflicts. But hopefully at this point it is mostly cleanup :)

**Comment by scopatz** [comment]
Friendly nudge again @dotsdl :wink:

**Comment by dotsdl** [comment]
Going to try to address merge conflicts today. Been a bit low-bandwidth lately.

**Comment by scopatz** [comment]
Thanks @dotsdl!

**Comment by dotsdl** [comment]
Sorry @scopatz, came down with the cold and been sick the past several days. Whittling down the merge conflicts. About 193 to go; going to see if I can't knock them out before too many commits land on master.

**Comment by scopatz** [comment]
No worries! I have started to resolved the conflicts but haven't finished yet. I caused them so I feel like I should clean them up :wink:

**Comment by dotsdl** [comment]
Thanks @scopatz! I wouldn't have had a chance to get to this until this weekend at the earliest.

**Comment by scopatz** [comment]
Alright! I think I have resolved all of the conflicts with master!

**Comment by dotsdl** [comment]
Still broken? Would you like to divide up the work?

**Comment by scopatz** [comment]
Hi @dotsdl, I have gotten the tests working! At this point, if you wanted to write up a news entry, that would be awesome.  Also, right now, if you try to access the old API, you get an attribute error:

```
scopatz@artemis ~ $ import builtins
scopatz@artemis ~ $ builtins.__xonsh_env__
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/home/scopatz/.local/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 16268, in default
    run_compiled_code(code, self.ctx, None, "single")
  File "/home/scopatz/.local/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 2448, in run_compiled_code
    func(code, glb, loc)
  File "<xonsh-code>", line 1, in <module>
AttributeError: module 'builtins' has no attribute '__xonsh_env__'
```

I think we should have proxies that redirect to the new API and issue a deprecation warning for the time being.

**Comment by scopatz** [comment]
Would you mind doing the proxy bit too?

**Comment by dotsdl** [comment]
I can do that. Thinking about implementation though, I'm not sure where to stick the deprecation warnings. Not sure it's possible to do this on the `builtins` module object as descriptors, but also not sure it's possible for the individual attributes except through perhaps their own `__getattr__` or `__getitem__` methods. Thoughts?

**Comment by scopatz** [comment]
Yeah, I don't think that this is doable via module-level descriptors (which I think are coming in Python v3.8? Maybe they are in v3.7?).  Rather, I think there needs to be something like a ProxyWarning object where `__getattr__`, `__setattr__`, and `__delattr__` are overwritten.  Something like:

```python
class ProxyWarning:
    def __init__(self, obj, badname, goodname):
        self.obj = obj
        self.badname = badname
        self.goodname = goodname

    def __getattr__(self, name):
        print("{badname} has been deprecated, please use {goodname} instead.")
        return getattr(self.obj, name)


builtins.__xonsh_env__ = ProxyWarning(
    builtins.__xonsh__.env, 
    "builtins.__xonsh_env__",
    " builtins.__xonsh__.env"
)
```

**Comment by dotsdl** [comment]
@scopatz, I think I managed to get everything in place. Watching tests. Included changes to account for updates to master in time since.

**Comment by dotsdl** [comment]
Ready for review.

**Comment by gforsyth** [comment]
Hey @dotsdl -- thanks for putting in so much work on this!  Especially in re: keeping up with a moving target.

2 initial things that I think need to be resolved:
1. The old `__` proxy doesn't seem to be intercepting old-style calls, I'm getting a bunch of `NameError: name '__xonsh_env__' is not defined`
2. Whatever deprecation message we give, we should probably also allow for a way of suppressing it (which, I know, defeats the point...) because this is going to break most, if not all, `xontribs` and it's going to take a while to get the API changes propagated out to all of them.

**Comment by gforsyth** [comment]
Oh, and please put in a news entry, too!

**Comment by dotsdl** [comment]
On it. Working out the kinks now.

**Comment by gforsyth** [comment]
Hey @dotsdl, looks like CI is failing due to some of the proxying behavior.  I can try to take a closer look later tonight.

**Comment by scopatz** [comment]
Thanks again @dotsdl! The tests now pass locally, let's keep our fingers crossed for CI

**Comment by scopatz** [comment]
Ok @gforsyth, I think this may be ready for a final review and merge, if you'd like to do the honors! (I feel I have committed too much to this branch to review)

**Comment by gforsyth** [comment]
I think there are a few more places that this needs to be updated (very possible I'm wrong but I thought there should be eyes on these):

```
ast.py
637:    """calls getattr(name, '__xonsh_block__', False)."""
642:            Str(s="__xonsh_block__", lineno=lineno, col_offset=col),

built_ins.py
704:        thable = getattr(alias, "__xonsh_threadable__", True)
709:        cpable = getattr(alias, "__xonsh_capturable__", self.captured)
1210:    kind = getattr(obj, "__xonsh_block__", str)

contexts.py
14:    __xonsh_block__ = str
80:        name = "__xonsh_functor_{uid}__".format(uid=uid)

tools.py
2251:    f.__xonsh_threadable__ = False
2260:    f.__xonsh_capturable__ = False
```

**Comment by scopatz** [comment]
So `__xonsh_block__` is [part of the API for context manager macros](https://xon.sh/tutorial_macros.html?highlight=__xonsh_block__#context-manager-macros)

`__xonsh_threadable__` and `__xonsh_capturable__` are part of an internal API for specifying whether callable aliases are threadable or capturable, eg see [unthreadable](https://xon.sh/api/tools.html?highlight=unthreadable#xonsh.tools.unthreadable)

So these names need not change because they don't live in `builtins`. But good catch :wink:

File: xonsh/built_ins.py [hunk 1]
```diff
@@ -1221,83 +1224,277 @@ def load_builtins(execer=None, ctx=None):
...
+
+
+class XonshSession:
+    """All components defining a xonsh session.
+
+    """
+
+    def __init__(self, execer=None, ctx=None):
+        """
+        Paramters
```
Comment (gforsyth) [inline xonsh/built_ins.py]: *Parameters

File: xonsh/built_ins.py [hunk 2]
```diff
@@ -1221,83 +1224,277 @@ def load_builtins(execer=None, ctx=None):
...
+        execer : Execer, optional
+            Xonsh execution object, may be None to start
+        ctx : Mapping, optional
+            Context to start xonsh session with.
+        """
+        self.execer = execer
+        self.ctx = {} if ctx is None else ctx
+
+    def load(self, execer=None, ctx=None):
+        """Loads the session with deafult values.
```
Comment (gforsyth) [inline xonsh/built_ins.py]: *default

File: xonsh/built_ins.py [hunk 3]
```diff
@@ -1221,83 +1224,277 @@ def load_builtins(execer=None, ctx=None):
...
+
+        Paramters
```
Comment (gforsyth) [inline xonsh/built_ins.py]: Parameters

**Comment by scopatz** [comment]
@gforsyth - Wow! Amazing eye! Fixes in.

**Comment by gforsyth** [comment]
Hrmmm... there are some failures on Appveyor that don't look like the normal failures.

**Comment by scopatz** [comment]
All the failures I see are the normal, random runtime integration test failures. The exception to this is the msys failures (last build), which are allowed.

**Comment by gforsyth** [comment]
I haven't seen this one before (py35) :

```
_____________________________ test_script[case0] ______________________________
case = ("\ndef _f():\n    print('hello')\n\naliases['f'] = _f\nf\n", 'hello\n', 0)
    @pytest.mark.parametrize("case", ALL_PLATFORMS)
    def test_script(case):
        script, exp_out, exp_rtn = case
        out, err, rtn = run_xonsh(script)
>       assert exp_out == out
E       assert 'hello\n' == 'hello\nC:\\projects\\xo...ls(), "RE_FSTR_ENVVAR"\n'
E           hello
E         + C:\projects\xonsh\xonsh\parsers\base.py:30: DeprecationWarning: invalid escape sequence \{
E         +   lambda: re.compile("\{\s*\$(\w+)"), globals(), "RE_FSTR_ENVVAR"
tests\test_integrations.py:348: AssertionError
============================== warnings summary ===============================
```

**Comment by scopatz** [comment]
Oh you are right, that's weird

**Comment by scopatz** [comment]
I think I have fixed this, but appveyor is taking forever to run

**Comment by scopatz** [comment]
This seems to have passed!

**Comment by gforsyth** [comment]
🎉 🎉 
Thanks @dotsdl and @scopatz!  It's great to have this in!

Now to re-write all the external xontribs!

**Comment by scopatz** [comment]
:tada:!

**Comment by dotsdl** [comment]
Awesome! Thanks for the support @scopatz and @gforsyth! Glad this is finally complete. 🍾

**Pull request merged**